### PR TITLE
server: sql listen before tenant id set

### PR DESCRIPTION
**Pull request by darinpp**
Previously, running SQL tenant processes required to specify the tenant id as command flag. That prevented starting an SQL tenant process before the tenant id associated is known and made it impossible to have a pool of generic warm pods waiting to be associated with a tenant. Not being able to have such pool makes also the cold start latency include the cost of starting a new SQL tenant process. It was also possible that a tenant connection is forwarded to a newly started tenant SQL process before its SQL listener is on so clients may get refused connections. To solve the problem, this PR adds a new flag that allows the tenant id to be set by reading a file after the process starts. Such file, will be empty when the SQL process starts and will be populated when the SQL process is to be associated with a specific tenant. When the tenant id is set through a file, and the file is empty at startup, the SQL process will block and wait for the tenant id to be written in the file. This is combined with a change where the SQL listener is created before the tenant id is read from the file. This eliminates the posibility that a client connection may be refused if it is forwarded quickly to the SQL process.

Epic: CC-8504

Release note: None

### CODE REVIEW

**Review by knz — DISMISSED** [review]
While I welcome the introduction of an "early listener" for the second part of the change, I feel the mechanism chosen here to feed the tenant ID into the process is yet fully thought out.

1. we generally eschew fsnotify or other "automatic" mechanisms to pick up env changes, and instead favor picking up new information using an explicit signal (for example, cert rotation occurs when the process is sent SIGHUP). The main reason is that fsnotify will pick up any write to the file, _even partial writes_ and so you could pick up a garbage/truncated tenant ID when receiving the fsnotify notifcation. The signal delivery doesn't have this problem.
2. when Jeff and I last discussed this, we were also considering that delivery using an API (HTTP or RPC) would result in better ergonomics for the CC control plane. Organizing the delivery of a file change inside the container seems more complex to implement and more brittle. What became of that plan?

> Reviewed 4 of 10 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-:-NdfqKeYB7fAMoTyV5hg:bhin2bt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @darinpp, @jaylim-crl, and @JeffSwenson)

___
*[`pkg/cli/flags.go` line 1306 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-Ndfq_NdEF3A8i4FKRSl:-Ndfq_NdEF3A8i4FKRSm:bdn71tf) ([raw file](https://github.com/cockroachdb/cockroach/blob/352040efb346a3bf17ef9a82eacdb9e739daa9fc/pkg/cli/flags.go#L1306)):*
> ```Go
> 		// We assume that we only need to change top level store as temp dir configs are
> 		// initialized when start is executed and temp dirs inherit path from first store.
> 		serverCfg.Stores.Specs[0].Path += "-tenant"
> ```

This is not a good change. We need to differentiate the paths here when folk are testing on their dev computer.
___
*[`pkg/server/tenant.go` line 260 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-NdfqmwF-_Y2b1RAaRZA:-NdfqmwF-_Y2b1RAaRZB:bum07gl) ([raw file](https://github.com/cockroachdb/cockroach/blob/352040efb346a3bf17ef9a82eacdb9e739daa9fc/pkg/server/tenant.go#L260)):*
> ```Go
> 	// (until the server is ready) won't cause client connections to be rejected.
> 	if baseCfg.SplitListenSQL && !baseCfg.DisableSQLListener {
> 		if baseCfg.SQLAddrListener, err = ListenAndUpdateAddrs(
> ```

pls don't assign to `baseCfg` if there's a non-nil err.

**Comment by darinpp** [comment]
> 1. we generally eschew fsnotify or other "automatic" mechanisms to pick up env changes, and instead favor picking up new information using an explicit signal (for example, cert rotation occurs when the process is sent SIGHUP). The main reason is that fsnotify will pick up any write to the file, _even partial writes_ and so you could pick up a garbage/truncated tenant ID when receiving the fsnotify notifcation. The signal delivery doesn't have this problem.
> 2. when Jeff and I last discussed this, we were also considering that delivery using an API (HTTP or RPC) would result in better ergonomics for the CC control plane. Organizing the delivery of a file change inside the container seems more complex to implement and more brittle. What became of that plan?
> 
Delivery through a push is a problem as we have to handle the case where cockroachdb crashes and is restarted. It would need to be able to find the tenant id in this case (as in read from a file) or do a request and get a reply back with it. Here are two suggestion:
- write the tenant id to a file and use SIGHUP (instead of fsnotify) to indicate readiness. On restart, cockroachdb will just read the file again and not block (and wait for SIGHUP). 
- instead of a file, use unix domain socket, make the cockroach db request the tenant id and the other side to reply when requested. 

Thoughts?

**Comment by knz** [comment]
>  write the tenant id to a file and use SIGHUP (instead of fsnotify) to indicate readiness. On restart, cockroachdb will just read the file again and not block (and wait for SIGHUP).

this would work, let's keep it as backup idea if the one below doesn't work out
 
>   instead of a file, use unix domain socket, make the cockroach db request the tenant id and the other side to reply when requested.

That would work but would be very hard to operate through containers. Wouldn't it be easier to fetch the tenant ID over a HTTP API instead?

**Comment by knz** [comment]
fetch the tenant ID *and the TLS certs* over that HTTP request maybe? 
IIRC we need different certs for each tenant ID value. @JeffSwenson wdyt?

**Review by darinpp — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110082)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jaylim-crl, @JeffSwenson, and @knz)

___
*[`pkg/cli/flags.go` line 1306 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-Ndfq_NdEF3A8i4FKRSl:-Ndg6bxt3xFbHwkh8k9C:b46sa69) ([raw file](https://github.com/cockroachdb/cockroach/blob/352040efb346a3bf17ef9a82eacdb9e739daa9fc/pkg/cli/flags.go#L1306)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

This is not a good change. We need to differentiate the paths here when folk are testing on their dev computer.
</blockquote></details>

My understanding is this simply changes the default. It doesn't prevent you to specify explicit path. Do you have a specific suggestion as to how to handle this?
___
*[`pkg/server/tenant.go` line 260 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-NdfqmwF-_Y2b1RAaRZA:-Ndg7wju5vqRiXfmw8i4:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/352040efb346a3bf17ef9a82eacdb9e739daa9fc/pkg/server/tenant.go#L260)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

pls don't assign to `baseCfg` if there's a non-nil err.
</blockquote></details>

fixed

**Review by darinpp — COMMENTED** [review]
So cockroachdb will do HTTP request (with some large timeout) and wait for the tenant id to be provided (at startup)? As the pods start not associated with a particular tenant and can stay like this for long time (hours) the response may came after some time. If a restart happens, the request will be answered immediately so there will not be any wait. Would that be OK?

The certs are not an issue. They can be loaded in the same way as they are loaded now for the time being. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-:-Ndg8JiR-hLCXxQ_idRw:bna66oy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jaylim-crl, @JeffSwenson, and @knz)

**Comment by jaylim-crl** [comment]
> fetch the tenant ID and the TLS certs over that HTTP request maybe?
IIRC we need different certs for each tenant ID value. @JeffSwenson wdyt?

If we do that, we should do it over HTTPS. This is the same idea as moving sqlstarter into the DB. The starter process today receives certs together with tenant ID in a zip file via an HTTPS request, and then runs execve(2) into the cockroach process. We also write a checkpoint file locally. If the process crashes, sqlstarter will detect that the process is ready to go, and swaps to the cockroach process right away without starting the HTTPS listener.

**Review by knz — COMMENTED** [review]
After discussion, it appears that the callback and delayed listener get us the best bang for the buck.
The callback in particular is interesting because later we can pull it even farther inside `(*SQLServerWrapper).PreStart`.

I still feel we're discussing whether the tenant ID is read from a file or using HTTP, but otherwise the structure of this PR feels fine to me.

> Reviewed 6 of 10 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-:-NdkI2yy32zFGyOll19k:b-pbc67o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @darinpp, @jaylim-crl, and @JeffSwenson)

___
*[`pkg/cli/flags.go` line 1306 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-Ndfq_NdEF3A8i4FKRSl:-NdkHwym5mRJ5FwTVlxz:b4hnq3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/352040efb346a3bf17ef9a82eacdb9e739daa9fc/pkg/cli/flags.go#L1306)):*
<details><summary><i>Previously, darinpp wrote…</i></summary><blockquote>

My understanding is this simply changes the default. It doesn't prevent you to specify explicit path. Do you have a specific suggestion as to how to handle this?
</blockquote></details>

add a random string suffix instead.
___
*[`pkg/cli/flags.go` line 550 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-NdkHQkF4ujgIus9OvlI:-NdkHQkF4ujgIus9OvlJ:b1bi7zn) ([raw file](https://github.com/cockroachdb/cockroach/blob/982984acbef79041d8a7624a94f16ebf36ff4c07/pkg/cli/flags.go#L550)):*
> ```Go
> 		f := mtStartSQLCmd.Flags()
> 		cliflagcfg.VarFlag(f, &tenantIDWrapper{&serverCfg.SQLConfig.TenantID}, cliflags.TenantID)
> 		cliflagcfg.VarFlag(f, &tenantIDFromFileWrapper{
> ```

I feel this is too complicated. Please consider having the flag point to a simple file name string,
then assign `DelayedSetTenantID` conditionally under `runStartInternal` (i.e `if flag := ..; flag.Changed() {  config.DelayedSetTenantID = func() { read(filename) } }`)

Then you don't need the wrapper struct.
___
*[`pkg/server/config.go` line 489 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-NdkH6sT6oMmFLi8mlrg:-NdkH6sT6oMmFLi8mlrh:b-y9zm7) ([raw file](https://github.com/cockroachdb/cockroach/blob/982984acbef79041d8a7624a94f16ebf36ff4c07/pkg/server/config.go#L489)):*
> ```Go
> 
> 	// If set, will to be called at server startup to set up the tenant id
> 	DelayedSetTenantID func(*roachpb.TenantID) error
> ```

Maybe make this _return_ the tenant ID instead of writing to it in-place.

**Comment by knz** [comment]
@darinpp do you want to try adding a signal wait in addition (or replacement) to fsnotify, and do the simplification i suggested in the review?

**Comment by darinpp** [comment]
> @darinpp do you want to try adding a signal wait in addition (or replacement) to fsnotify, and do the simplification i suggested in the review?

Yes, I'll update the PR.

**Comment by darinpp** [comment]
> @darinpp do you want to try adding a signal wait in addition (or replacement) to fsnotify, and do the simplification i suggested in the review?

Having the signal wait in addition to fsnotify won’t have much effect. If the file doesn’t have the tenant id then fsnotify will block. Once the file is updated, the fsnotify will trigger and the code waiting for the tenant id will read and continue. The signal doesn’t matter. If the file has the tenant id then neither the signal nor the fsnotify are needed.

Making the sql server sidecar send a signal will require some way to find the PID of the sql server given that the sql server sidecar and the sql server are two different containers. I can do that by checking what processes are running but that is not perfect. Although this is not our use case, it is possible to have multiple cockroach images running and then it wouldn’t be clear which one should SIGHUP be sent to. Or it is possible that these are namespaced (again not our use case) and the process responsible for sending SIGHUP can’t enumerate the running processes in a namespace that isn’t its own.

The SIGHUP would work nicely if the sql server sidecar is the one that starts the sql server as a child process. Then it would have the PID and would be able to do the signaling easy. But the decision to have it run separate is to give k8s ability to constrain resources on the sql server (memory, cpu etc)

**Comment by knz** [comment]
Thanks that works.

I'll let you simplify the flag code before final review.

**Review by darinpp — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110082)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jaylim-crl, @JeffSwenson, and @knz)

___
*[`pkg/cli/flags.go` line 1306 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-Ndfq_NdEF3A8i4FKRSl:-NdnGnr48is6W1gvw6bs:bd3k3r2) ([raw file](https://github.com/cockroachdb/cockroach/blob/352040efb346a3bf17ef9a82eacdb9e739daa9fc/pkg/cli/flags.go#L1306)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

add a random string suffix instead.
</blockquote></details>

added the process PID which is a tad better than random string as it can help with matching a process with a store path
___
*[`pkg/cli/flags.go` line 550 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-NdkHQkF4ujgIus9OvlI:-NdnGk8Q7n1D7THGND3Y:b-4yfvz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/982984acbef79041d8a7624a94f16ebf36ff4c07/pkg/cli/flags.go#L550)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I feel this is too complicated. Please consider having the flag point to a simple file name string,
then assign `DelayedSetTenantID` conditionally under `runStartInternal` (i.e `if flag := ..; flag.Changed() {  config.DelayedSetTenantID = func() { read(filename) } }`)

Then you don't need the wrapper struct.
</blockquote></details>

refactored and removed the wrapper
___
*[`pkg/server/config.go` line 489 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-NdkH6sT6oMmFLi8mlrg:-NdnH0XZ1D7EbYqpszPh:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/982984acbef79041d8a7624a94f16ebf36ff4c07/pkg/server/config.go#L489)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Maybe make this _return_ the tenant ID instead of writing to it in-place.
</blockquote></details>

done

File: pkg/cli/flags.go [hunk 1]
```diff
@@ -946,29 +952,105 @@ func init() {
...
+// line of the file that should contain the `<TenantID>`.
+func tenantIDFromFile(
+	fileName string, watcherWaitCount *atomic.Uint32, watcherEventCount *atomic.Uint32,
+) (*roachpb.TenantID, error) {
+
+	// Start watching the file for changes as the typical case is that the file
+	// will not have yet the tenant id at startup.
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
```
Comment (JeffSwenson) [inline pkg/cli/flags.go:998]: nit: constructing the watcher should annotate the error and all of the errors returned by the watcher should be annotated. From experience, its common for CI environments and dev machines to run out of inotify handles. The kernel errors are vague and difficult to root cause.

File: pkg/cli/flags.go [hunk 2]
```diff
@@ -946,29 +952,105 @@ func init() {
...
+	}
+	defer func() { _ = watcher.Close() }()
+	if err = watcher.Add(fileName); err != nil {
+		return nil, err
+	}
+
+	headBuf := make([]byte, 100)
+	for {
+		readCount, err := func() (int, error) {
+			file, err := os.Open(fileName)
```
Comment (JeffSwenson) [inline pkg/cli/flags.go:1008]: nit: you can simplify this a little using os.ReadFile. It handles opening and closing the file.

```go
headBuf, err := os.ReadFile(filename)
if err != nil {
   return err
}
```

File: pkg/cli/flags.go [hunk 3]
```diff
@@ -946,29 +952,105 @@ func init() {
...
+			if err != nil {
+				return 0, err
+			}
+			defer file.Close()
+			return io.ReadFull(file, headBuf)
+		}()
+		if err != nil && !errors.IsAny(err, io.EOF, io.ErrUnexpectedEOF) {
```
Comment (JeffSwenson) [inline pkg/cli/flags.go:1015]: Do we actually expect EOF errors here or is this a hold over from when this was implemented via a pipe?

**Review by knz — COMMENTED** [review]
> Reviewed 5 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110082)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @darinpp and @jaylim-crl)

___
*[`pkg/cli/flags.go` line 550 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-NdkHQkF4ujgIus9OvlI:-NdpC0FB55bUtcMU_pps:b-f8zng4) ([raw file](https://github.com/cockroachdb/cockroach/blob/982984acbef79041d8a7624a94f16ebf36ff4c07/pkg/cli/flags.go#L550)):*
<details><summary><i>Previously, darinpp wrote…</i></summary><blockquote>

refactored and removed the wrapper
</blockquote></details>

Ok. Don't forget to reset the global variable in function `initPreFlagsDefaults()`.
___
*[`pkg/cli/flags.go` line 993 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-NdpCbLj0kPcW4s0TysS:-NdpCbLj0kPcW4s0TysT:b3dc7jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e6239f39420b3c5af41076e4c24072c4f891312/pkg/cli/flags.go#L993)):*
> ```Go
> 	fileName string, watcherWaitCount *atomic.Uint32, watcherEventCount *atomic.Uint32,
> ) (*roachpb.TenantID, error) {
> ```

nit: remove empty line
___
*[`pkg/cli/flags.go` line 1026 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-NdpCmnFEn9U7me1IrW7:-NdpCmnFEn9U7me1IrW8:b9k75oh) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e6239f39420b3c5af41076e4c24072c4f891312/pkg/cli/flags.go#L1026)):*
> ```Go
> 		}
> 		if readCount == len(headBuf) {
> 			// no new line and the file is larger than the initial portion that is
> ```

Here and below: comments contain sentences that start with a capital and end with a period.
___
*[`pkg/server/config.go` line 489 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-NdkH6sT6oMmFLi8mlrg:-NdpCzSQBver-V5JwBCm:b-j4kr5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/982984acbef79041d8a7624a94f16ebf36ff4c07/pkg/server/config.go#L489)):*
<details><summary><i>Previously, darinpp wrote…</i></summary><blockquote>

done
</blockquote></details>

The first return value does not need to be a pointer.

**Review by darinpp — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110082)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jaylim-crl, @JeffSwenson, and @knz)

___
*[`pkg/cli/flags.go` line 550 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-NdkHQkF4ujgIus9OvlI:-NdqjZQu4Vi9o21vBGjq:b-7tfmjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/982984acbef79041d8a7624a94f16ebf36ff4c07/pkg/cli/flags.go#L550)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Ok. Don't forget to reset the global variable in function `initPreFlagsDefaults()`.
</blockquote></details>

added
___
*[`pkg/cli/flags.go` line 982 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110082#gh-1319904032:-Ndqjbio0ZdAJPZHONJM:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e6239f39420b3c5af41076e4c24072c4f891312/pkg/cli/flags.go#L982)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

nit: remove the dead code in this comment.
</blockquote></details>

done
___
*[`pkg/cli/flags.go` line 993 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-NdpCbLj0kPcW4s0TysS:-NdqjcnO5hBPt9moTi6h:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e6239f39420b3c5af41076e4c24072c4f891312/pkg/cli/flags.go#L993)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

nit: remove empty line
</blockquote></details>

removed
___
*[`pkg/cli/flags.go` line 998 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110082#gh-1319913636:-Ndqje_y-Gsi3jvHOE8O:b3w6vks) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e6239f39420b3c5af41076e4c24072c4f891312/pkg/cli/flags.go#L998)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

nit: constructing the watcher should annotate the error and all of the errors returned by the watcher should be annotated. From experience, its common for CI environments and dev machines to run out of inotify handles. The kernel errors are vague and difficult to root cause. 
</blockquote></details>

added more details
___
*[`pkg/cli/flags.go` line 1008 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110082#gh-1319923549:-Ndqjuvc7KsaSiXV4AXX:bnz6psf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e6239f39420b3c5af41076e4c24072c4f891312/pkg/cli/flags.go#L1008)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

nit: you can simplify this a little using os.ReadFile. It handles opening and closing the file.

```go
headBuf, err := os.ReadFile(filename)
if err != nil {
   return err
}
```
</blockquote></details>

switched to ReadFile
___
*[`pkg/cli/flags.go` line 1015 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110082#gh-1319924500:-Ndqjz1ABMTXJZ2wt-GS:bunx4p1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e6239f39420b3c5af41076e4c24072c4f891312/pkg/cli/flags.go#L1015)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

Do we actually expect EOF errors here or is this a hold over from when this was implemented via a pipe?
</blockquote></details>

not an issue anymore with ReadFile
___
*[`pkg/cli/flags.go` line 1026 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-NdpCmnFEn9U7me1IrW7:-NdqkNIZBT3IqHbMj-lz:b-uo6fys) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e6239f39420b3c5af41076e4c24072c4f891312/pkg/cli/flags.go#L1026)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Here and below: comments contain sentences that start with a capital and end with a period.
</blockquote></details>

All comments now fixed.
___
*[`pkg/cli/cliflags/flags_mt.go` line 22 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110082#gh-1319909911:-NdqkSGY2YnUSbS9_ZfC:b-wui1lx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e6239f39420b3c5af41076e4c24072c4f891312/pkg/cli/cliflags/flags_mt.go#L22)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

What do you think of naming this tenant-id-file? I think noun flags tend to be a little easier to comprehend than verb flags. It also matches the other mt flags like "allowlist-file" and "denylist-file".
</blockquote></details>

I renamed it to `tenant-id-file`
___
*[`pkg/server/config.go` line 489 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-NdkH6sT6oMmFLi8mlrg:-NdqkYvx0CRNjJahNbhq:bpr4zhk) ([raw file](https://github.com/cockroachdb/cockroach/blob/982984acbef79041d8a7624a94f16ebf36ff4c07/pkg/server/config.go#L489)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

The first return value does not need to be a pointer.
</blockquote></details>

I changed it to not be a pointer.

**Review by knz — COMMENTED** [review]
Great. I feel this is nearing completion.

> Reviewed 5 of 7 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-:-Ne34Be57f4NsEJBUleX:b-rfhq4x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @darinpp, @jaylim-crl, and @JeffSwenson)

___
*[`pkg/cli/flags.go` line 964 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-Ne33qTVBU67q4zAWCQR:-Ne33qTVBU67q4zAWCQS:bhua92t) ([raw file](https://github.com/cockroachdb/cockroach/blob/23c9ad926d022a89fe90a779e7bf26ff5a5f81cd/pkg/cli/flags.go#L964)):*
> ```Go
> 		return roachpb.TenantID{}, errors.New("invalid tenant ID")
> 	}
> 	return roachpb.MustMakeTenantID(tenID), nil
> ```

nit: roachpb.MakeTenantID and return the error (there's more checks there).
___
*[`pkg/cli/cliflags/flags_mt.go` line 23 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-Ne343Rg0zeeQniqLyk-:-Ne343Rg0zeeQniqLyk0:b-ppm633) ([raw file](https://github.com/cockroachdb/cockroach/blob/23c9ad926d022a89fe90a779e7bf26ff5a5f81cd/pkg/cli/cliflags/flags_mt.go#L23)):*
> ```Go
> 	TenantIDFile = FlagInfo{
> 		Name:   "tenant-id-file",
> 		EnvVar: "COCKROACH_TENANT_ID_FILE",
> ```

Do we truly need an env var? (Maybe more a question for jeff/jay?)

**Review by jeffswenson — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110082)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @darinpp, @jaylim-crl, and @knz)

___
*[`pkg/cli/cliflags/flags_mt.go` line 23 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-Ne343Rg0zeeQniqLyk-:-Ne3MZ0r3ArbY1CsQ8aC:b-3sv3o0) ([raw file](https://github.com/cockroachdb/cockroach/blob/23c9ad926d022a89fe90a779e7bf26ff5a5f81cd/pkg/cli/cliflags/flags_mt.go#L23)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Do we truly need an env var? (Maybe more a question for jeff/jay?)
</blockquote></details>

I don't think we need the environment variable. One case where environment variables can be helpful is it allows us to set an option before the binary supports the option. That doesn't really apply here because the setting tenant-id-file changes the protocol between the sql starter and the sql server, so the environment variable's forwards compatibility doesn't really help.

**Review by jeffswenson — APPROVED** [review]
LGTM

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-:-Ne3NpSODc7A-Tf-67JZ:bb6ywmx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @darinpp, @jaylim-crl, and @knz)

**Review by darinpp — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110082)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jaylim-crl, @JeffSwenson, and @knz)

___
*[`pkg/cli/flags.go` line 964 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-Ne33qTVBU67q4zAWCQR:-Ne4HQ6c1CmbUAD1ORTv:b-sf24fm) ([raw file](https://github.com/cockroachdb/cockroach/blob/23c9ad926d022a89fe90a779e7bf26ff5a5f81cd/pkg/cli/flags.go#L964)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

nit: roachpb.MakeTenantID and return the error (there's more checks there).
</blockquote></details>

Done. Also removed the `TenantID == 0` check as this is what `MakeTenantID` does.
___
*[`pkg/cli/cliflags/flags_mt.go` line 23 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110082#-Ne343Rg0zeeQniqLyk-:-Ne4HM8YBKBl182CPR-I:bx922lj) ([raw file](https://github.com/cockroachdb/cockroach/blob/23c9ad926d022a89fe90a779e7bf26ff5a5f81cd/pkg/cli/cliflags/flags_mt.go#L23)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

I don't think we need the environment variable. One case where environment variables can be helpful is it allows us to set an option before the binary supports the option. That doesn't really apply here because the setting tenant-id-file changes the protocol between the sql starter and the sql server, so the environment variable's forwards compatibility doesn't really help.
</blockquote></details>

I removed it

**Review by knz — APPROVED** [review]
> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110082)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jaylim-crl and @JeffSwenson)

**Pull request merged**